shotover/src/codec/cassandra.rs

```diff
@@ -561,46 +659,107 @@ impl CassandraEncoder {
+                    Compression::None => {
+                        // write envelope header with dummy values for those we cant calculate till after we write the message
+                        let header_start = dst.len();
 
-                dst.extend_from_slice(&[0, 0, 0, 0, 0, 0]);
-                let payload_start = dst.len();
+                        dst.extend_from_slice(&[0, 0, 0, 0, 0, 0]);
+                        let payload_start = dst.len();
 
-                self.encode_envelope(dst, m)?;
+                        self.encode_envelope(dst, m, Compression::None)?;
 
-                //measure length of message and calculate crc24 and overwrite frame header values
-                let mut payload_len = dst.len() - payload_start;
+                        //measure length of message and calculate crc24 and overwrite frame header values
+                        let mut payload_len = (dst.len() - payload_start) as u64;
 
-                if true {
-                    // TODO if self_contained
-                    payload_len |= 1 << 17;
-                }
+                        if true {
+                            // TODO if self_contained
+                            payload_len |= 1 << 17;
+                        }
 
-                // add header length & header crc
-                let payload_len = &payload_len.to_le_bytes()[..3];
-                dst[header_start..header_start + 3].copy_from_slice(payload_len);
-                dst[header_start + 3..header_start + 6].copy_from_slice(
-                    &cassandra_protocol::crc::crc24(payload_len).to_le_bytes()[..3],
-                );
+                        // add header length & header crc
+                        let payload_len = &payload_len.to_le_bytes()[..3];
+                        dst[header_start..header_start + 3].copy_from_slice(payload_len);
+                        dst[header_start + 3..header_start + 6].copy_from_slice(
+                            &cassandra_protocol::crc::crc24(payload_len).to_le_bytes()[..3],
+                        );
 
-                // add payload crc
-                dst.extend_from_slice(
-                    &cassandra_protocol::crc::crc32(&dst[payload_start..]).to_le_bytes(),
-                );
+                        // add payload crc
+                        dst.extend_from_slice(
+                            &cassandra_protocol::crc::crc32(&dst[payload_start..]).to_le_bytes(),
+                        );
+                    }
+                    Compression::Lz4 => {
+                        let header_start = dst.len();
+
+                        dst.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]);
+                        let payload_start = dst.len();
+
+                        // TODO we should not be encoding small frames
+                        let uncompressed_len = self.encode_compressed_payload(dst, m)?;
+                        let compressed_len = dst.len() - payload_start;
```

Comment:
we should return an error when compressed_len or uncompressed_len are longer than supported by the protocol.
Trying to debug that case would be a nightmare because the bits would overlap, an error here would save us from that possibility.

Reply:
In that case we should be sending the envelope as a non-self contained frame

Reply:
thats fair, we should still add at least a:
```rust
if uncompressed_len < MAX_UNCOMPRESSED_LEN {
    todo!("non self contained frames not yet implemented.")
}
```

so that we dont run the risk of mangling the bits